Use task group for collecting tasks with the same lifetime

cowboyd opened the pull request:
### Motivation
Instead of using an option 'blockParent'. We can explicitly spawn different tasks into collections of groups that can be waited upon individually. This not only makes which group a task is part of explicit, but it allows flexibility in how you consume the blocking tasks.

This is fairly common in other structured concurrency frameworks like Swift or ember-concurrency, and can be expanded upon to add other aspects like "how many concurrent things can be going at the same time" or "what do we do when the max concurrency is attained... do we drop incoming operations? do we respond with a "503 operation"? do we cancel the currently running things? do we enqueue the new operations? How many do we enqueue?"

So this is an experiment in what that might look like.

cowboyd committed:
Use task group for collecting tasks with the same lifetime

Instead of using an option 'blockParent'. We can explicitly spawn
different tasks into collections of groups that can be waited upon
individually. This not only makes which group a task is part of
explicit, but it allows flexibility in how you consume the blocking
tasks.

This is fairly common in other structured concurrency frameworks like
Swift or ember-concurrency, and can be expanded upon to add other
aspects like "how many concurrent things can be going at the same
time" or "what do we do when the max concurrency is attained... do we
drop running tasks? do we cancel the currently running things? do we
enqueue the new operations?"

So this is an experiment in what that might look like.

> **Implementation Note**
> the task group itself is a resource so that you can `yield` to
> it. That's why it has a resource that returns a resource. That's
> kinda cheating because really we want the ability for any object to
> be an operation.

cowboyd commented on packages/server/src/task-group.ts:
```diff
@@ -0,0 +1,38 @@
+            let t: Task<T> = task = yield scope.spawn(operation);
+            tasks.add(t);
+            try {
+              yield t.future;
```
We don't care how this task ended, only that it did. Since it was spawned in its parent, it will fail it if needs be.

cowboyd commented on packages/server/src/task-group.ts:
```diff
@@ -0,0 +1,38 @@
+          return task;
+        },
+        *init() {
+          for (let task of tasks) {
+            yield task.future;
+          }
```
Probably need to wrap this in a `while (!!tasks.size) { }` in case new tasks are added to the group while it is shutting down.

dagda1 commented on packages/server/src/task-group.ts:
```diff
@@ -0,0 +1,38 @@
+        *spawn<T>(operation: Operation<T>) {
+          let task: Task<T> | undefined;
+          yield spawn(function*() {
+            let t: Task<T> = task = yield scope.spawn(operation);
```
why is there a local `t` and `task` var?

cowboyd replied:
Excellent point! It was mainly there to help with the typing because the type of `task` has to be `Task<T> | undefined` and I can't add it to the set or delete it even if it's undefined. In truth though, it is probably better not to have two storage locations in memory when one will suffice.

dagda1 commented:
Looks like a good idea

Should this not be part of effection itself?

At the moment, I don't see any max task restriction, is this still to be added?

cowboyd committed:
👍Only use one variable for a task group

cowboyd commented:
> Looks like a good idea
> 
> Should this not be part of effection itself?
> 
> At the moment, I don't see any max task restriction, is this still to be added?

Yes, this is more of a proving ground for the idea, and then when we're happy with it we can add it to Effection proper. I was thinking something like

```ts
function* run() {
  let group = yield createTaskGroup('socket handlers', {
    maxConcurrency: 200,
    drop: 'latest'
  });
}
```

cowboyd commented:
@dagda1 Yeah, this is something that I think should eventually be added to Effection itself, but that I was using BigTest as a testing ground.  I'd imagine it would be something like:

```ts
let group = yield createTaskGroup('socket handlers', {
  maxConcurrency: 10,
  drop: 'latest'
});
```

cowboyd committed:
Don't be clever about resources

Using a resource so that you could yield to the `TaskGroup` was too
clever by half, and it will also be obsolete once we have
`Symbol.operation` when
https://github.com/thefrontside/effection/pull/600 lands.

At that point we can just have a `[Symbol.operation]` that delegates
to `allSettled()`

For the mean time, this just uses an `allSettled` method which blocks
until all tasks in the group have completed for good or for ill. This
follows the convention in the wild, and currently in stage 4 of
promises https://github.com/tc39/proposal-promise-allSettled

dagda1 commented on packages/server/src/command-server.ts:
```diff
@@ -122,7 +122,7 @@ function handleSocketConnection(options: CommandServerOptions): (socket: Socket<
 
   return (socket) => function*() {
-    let handlers: TaskGroup = yield createTaskGroup('message handlers');
+    let handlers = yield* createTaskGroup('message handlers');
```
💯

cowboyd committed:
fix compile step
